Get subscriptions

trashstackdev opened the pull request:
# Pull Request

## Trello Ticket

- [Trello Ticket Link](insert link here)
  <!-- Link to the corresponding Trello ticket for better context -->

## Description

1. Erstellung von GET subscription, basierend auf den query parameter `type`
2. Auslagerung von "Subscribe" und "Unsubscribe" in den eigenen subscripion Handler
3. Auslagerung der parse pagination und send pagination in das `utils` package

## Additional Remarks

<!-- Any additional comments or context that would be helpful for the reviewer. -->

trashstackdev committed:
Add subscription handler and pagination utility

A new subscription handler has been added in this commit, which handles fetching followers and following subscriptions. For better code reuse, pagination-related functions which were originally in user_handler have been moved to a dedicated pagination utility file under utils. These files include parsing pagination parameters and sending paginated responses. These changes enhance modularity and help to avoid duplication. The requisite updates in user_handler, router, and dto schema have also been carried out.

trashstackdev committed:
Remove subscription methods from UserHandler

This commit involves removing the Subscribe and Unsubscribe methods from the UserHandler. Consequently, these operations have been transferred to SubscriptionHandler to improve code organization and avoid confusing handlers. Usage of relevant methods in router.go has also been updated to reflect these changes.

krieger2501 commented on internal/handlers/subscription_handler.go:
```diff
@@ -0,0 +1,282 @@
+
+type SubscriptionHandler struct {
+	DatabaseManager managers.DatabaseMgr
+	JwtManager      managers.JWTMgr
```
Den brauchen wir doch gar nicht hier, oder?

krieger2501 commented on internal/handlers/subscription_handler.go:
```diff
@@ -0,0 +1,282 @@
+
+// HandleGetSubscriptions retrieves the subscriptions of a user and sends a paginated response.
+func (handler *SubscriptionHandler) HandleGetSubscriptions(w http.ResponseWriter, r *http.Request) {
+	ctx, cancel := context.WithDeadline(r.Context(), time.Now().Add(1000*time.Second))
```
hast deine debug-zeit vergessen

krieger2501 commented on internal/handlers/subscription_handler.go:
```diff
@@ -0,0 +1,282 @@
+	results := make([]any, 0, limit)
+	var totalResults int
+	// read from query parameter
+	subscriptionType := r.URL.Query().Get("type")
```
hier kannst du noch einen `SubscriptionTypeParamKey` machen, oder?

krieger2501 commented on internal/schemas/dto_schema.go:
```diff
@@ -158,3 +158,15 @@ type ChangePasswordRequest struct {
+
+type FollowingDTO struct {
+	SubscriptionId uuid.UUID `json:"subscriptionId"`
+	CreationDate   string    `json:"creationDate"`
```
```suggestion
	SubscriptionDate string    `json:"subscriptionDate"`
```

siehe postman

krieger2501 commented on internal/schemas/dto_schema.go:
```diff
@@ -158,3 +158,15 @@ type ChangePasswordRequest struct {
+type FollowingDTO struct {
+	SubscriptionId uuid.UUID `json:"subscriptionId"`
+	CreationDate   string    `json:"creationDate"`
+	Following      AuthorDTO `json:"following"`
```
```suggestion
	User AuthorDTO `json:"user"`
```

Finde ich persönlich schöner mit Following, aber du hattest heute User geschrieben. Würde das aber eher in Postman ändern mit Einstimmung der anderen

krieger2501 reviewed (approved):
Sieht gut aus, meine Kommentare sind größtenteils nits.

Einzige Anmerkung ist, dass du relativ viel Code Duplication hast, weil sich die Queries ähneln. Wenn du Zeit und Lust hast, kannst du das ähnlich wie bei mir kürzen, ansonsten lagern wir das in ein späteres Refactoring aus

trashstackdev committed:
Refactor subscription handling and reduce context deadline

This commit simplifies subscription handling by removing the JwtManager from the NewSubscriptionHandler function and renaming various entities related to following and followers to a more generic SubscriptionUserDTO. The deadline for context in the HandleGetSubscriptions was also reduced from 1000 seconds to 10 seconds for better performance. A new constant 'SubscriptionTypeParamKey' was added in url_keys.go for improved code clarity.

trashstackdev committed:
Refactor subscription handler

The subscription handler has been refactored to streamline and simplify the handling of subscription types, using a more generic and dynamic approach while introducing a clearer distinction between 'subscribers' and 'subscribees'. Furthermore, 'SubscriptionTypeParamKey' constant has been renamed to 'type'.
